packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,77 @@
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const CWD = process.cwd()
+const file = (): string => CWD + `/.cache/redux.state`
+const chunkFilePrefix = (): string => CWD + `/.cache/redux.node.state_`
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+export const readFromCache = (): ICachedReduxState => {
+  // The cache is stored in two steps; the nodes in chunks and the rest
+  // First we revive the rest, then we inject the nodes into that obj (if any)
+  // Each chunk is stored in its own file, this circumvents max buffer lengths
+  // for sites with a _lot_ of content. Since all nodes go into a Map, the order
+  // of reading them is not relevant.
+
+  const obj: ICachedReduxState = v8.deserialize(readFileSync(file()))
+
+  // Note: at 1M pages, this will be 1M/chunkSize chunks (ie. 1m/10k=100)
+  const chunks = globSync(chunkFilePrefix() + `*`).map(file =>
+    v8.deserialize(readFileSync(file))
+  )
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
+
+  if (chunks.length) {
+    obj.nodes = new Map(nodes)
+  }
+
+  return obj
+}
+
+function guessSafeChunkSize(values: [string, IReduxNode][]): number {
+  const valueCount = values.length
+
+  // Pick a few random elements and measure their size.
+  // Pick a chunk size ceiling based on the worst case.
+  // This attempts to prevent small sites with very large pages from OOMing
+  let maxSize = 0
+  for (let i = 0; i < valueCount; i += Math.floor(valueCount / 11)) {
+    const size = v8.serialize(values[i]).length
+    maxSize = Math.max(size, maxSize)
+  }
+
+  // Max size of a Buffer is 2gb (yeah, we're assuming 64bit system)
+  // https://stackoverflow.com/questions/8974375/whats-the-maximum-size-of-a-node-js-buffer
+  // Use 1.5gb as the target ceiling, allowing for some margin of error
+  return Math.floor((150 * 1024 * 1024 * 1024) / maxSize)
+}
 
 export const writeToCache = (contents: ICachedReduxState): void => {
+  // Remove the old node files first, we may have fewer nodes than before and
+  // must make sure that no excess files are kept around.
+  globSync(chunkFilePrefix() + `*`).forEach(file => unlinkSync(file))
+
+  // Temporarily save the nodes and remove them from the main redux store
+  // This prevents an OOM when the page nodes collectively contain to much data
+  const map = contents.nodes
+  contents.nodes = undefined
   writeFileSync(file(), v8.serialize(contents))
+  // Now restore them on the redux store
+  contents.nodes = map
+
+  if (map) {
+    // Now store the nodes separately, chunk size determined by a heuristic
+    const values: [string, IReduxNode][] = [...map.entries()]
+    const chunkSize = guessSafeChunkSize(values)
+    const chunks = Math.ceil(values.length / chunkSize)
+
+    for (let i = 0; i < chunks; ++i) {
+      writeFileSync(
+        `${chunkFilePrefix()}${i}`,
+        v8.serialize(values.slice(i * chunkSize, i * chunkSize + chunkSize))
+      )
+    }
+  }
```

Comment:
this is cool 🎉

Reply:
With state split across many files we "lose" transactional (implicit) nature of using single file - we potentially can save subset of state now (I think we have try/catch somewhere higher in call stack that would make failure to save non-fatal potentially).

It would be good (IMO) to save files to tmp location and if all of them were successful move things around to final one?

Reply:
Oh yes, I forgot about the transactional state. We (or somebody else?) had a discussion about that. Yeah I'll use a tmp location instead, rename the current folder away, and the tmp folder in, and then destroy the old folder. That should be transactional enough... For now anyways. Thanks for reminding me!

Reply:
Swapping in/out a folder means backwards compat breaking.

To mitigate this I'll support the old file and a new folder. We can drop support for the legacy file when we bump a major. Just don't want any unexpected problems related to switching patch, or even minors.

Reply:
I think we might blow away entire cache when we upgrade gatsby version (not sure if we catch gatsby core version changes or just plugin version changes when we chose to invalidate cache). So we might not need to do this (needs verification)

Reply:
https://github.com/gatsbyjs/gatsby/blob/27aeb08f5d9bc055245b07247838308e1e404e26/packages/gatsby/src/bootstrap/index.js#L214-L231

Ok, we blow away cache if package.json changed - so this doesn't necesairly will catch cases of updating gatsby core (i.e. just blowing away lock files and reinstalling deps)

Reply:
Arguably we should. We could get the current running gatsby version and bake it into the cache and always invalidate it on a miss-match.

Reply:
/agree! (but probably in separate pull request)

Reply:
Probably ;)

Let's wait for green and, if you're alright with the latest changes, merge